packages/apollo-server-cache-redis/src/RedisClusterCache.ts

```diff
@@ -1,64 +1,19 @@
-import { KeyValueCache, KeyValueCacheSetOptions } from 'apollo-server-caching';
 import Redis, {
   ClusterOptions,
   ClusterNode,
   Redis as RedisInstance,
 } from 'ioredis';
-import DataLoader from 'dataloader';
+import { BaseRedisCache } from './BaseRedisCache';
 
-export class RedisClusterCache implements KeyValueCache {
+export class RedisClusterCache extends BaseRedisCache {
   readonly client: any;
-  readonly defaultSetOptions: KeyValueCacheSetOptions = {
-    ttl: 300,
-  };
-
-  private loader: DataLoader<string, string | null>;
 
   constructor(nodes: ClusterNode[], options?: ClusterOptions) {
-    const client = this.client = new Redis.Cluster(nodes, options);
-
-    this.loader = new DataLoader(
-      (keys = []) =>
-        Promise.all(keys.map(key => client.get(key).catch(() => null))),
```

Comment:
I've committed a potential solution, that fixes it in the injected client rather than adding logic to work around it in BaseRedisClient class, let me know what you think.

Reply:
On second thought after merging this means you can't actually create your own Redis.Cluster object, so I'd rather move the logic back into the base class.

Reply:
To me it would make sense if you wanted to create your own cluster cache to use the base class, the logic in the current RedisClusterCache seems mainly to work around ioredis, if you were using node-redis with the redis-clustr wrapper for example that already has logic to split multi-key commands. Having the base class allows apollo to just focus on core redis logic and let the consumer of the library handle anything implementation specific.

Reply:
Makes sense. I was trying to write the docs for #5088 to show examples and I thought it would be nice to be "all in" on the new class rather than showing the older pass-through implementations, but you can't really do that with `Redis.Cluster`. Thus the API change in #5088.

But I see what you mean about other implementations that do support mget and cluster.

What if I kept the structure from #5088 but changed the names to be `client` and `getOnlyClient` or `noMgetClient` or something?

Reply:
@adamnoakes Makes sense. I was trying to write the docs for #5088 to show examples and I thought it would be nice to be "all in" on the new class rather than showing the older pass-through implementations, but you can't really do that with `Redis.Cluster`. Thus the API change in #5088.

But I see what you mean about other implementations that do support mget and cluster.

What if I kept the structure from #5088 but changed the names to be `client` and `getOnlyClient` or `noMgetClient` or something?

Reply:
Hmm yeah i see, could do it via a flag? like `useMGet` that defaults to true. Would be cleaner imo than getOnlyClient/ noMgetClient as the client will likely still have mget (if they are just passing cluster ioredis client in as is for example). 

On the topic of going all in on the new class. Would be nice to rename it to just RedisCache when you do the next major/ breaking change release and remove the existing RedisCache/ RedisClusterCache classes as i think they are redundant with this approach.

Reply:
The flag seems like a nice idea, but I do like the type checking we get by providing two different arguments. I am going to go with noMgetClient though to make it clear that it's about what the client supports, not whether you're in a cluster. See https://github.com/apollographql/apollo-server/pull/5094/commits/21f148eb1e44c4dbd243d40959780f78b7b4c87d

Tracking the major version bump suggestion at https://github.com/apollographql/apollo-server/issues/5099

Reply:
Sounds good, thank you for your help on getting this merged.

On the flag type checking, I was curious so had a play and looks like you _could_ have a flag and type checking using overloads.

[tsplayground](https://www.typescriptlang.org/play?ssl=19&ssc=45&pln=1&pc=1#code/JYOwLgpgTgZghgYwgAgOITAYQDbAuZAbwChkzkBzDALmQAoBKZAXgD5kA3Ae2ABMBuYgF9ixUJFiIUAWXRZc+MEVLkAtlTC1GLdtz6CRxGAFcQCMMC4hkkAM5g6CBeFqyMOPOAA0yLgAcLK1taQmRjWwg3TRsoYxQhBlo9ASNTc0trOwcnT2i5D0Uff0CQYKIwiKjaeGwI5ASknhSTMxKbCHtHZzz3buQAH2Qogu9fAIyy0PDIuVoAIy4ubAg4awbOJrIScmQoDGMoa1NeCBhQCBTDLLpQjS0mNiIhIR8pytnkGrqEwWvQ9Ro9Ae7EIz1eFRmgLAsXiDEExD+lEB2keoJe5WmVRicXqcIRHQc-zuQJ0T3Rb0h0S+sP4QA)

```typescript
interface GetClient {
    get: () => void;
}

interface MGetClient {
    mget: () => void;
}

function test(client: MGetClient, options?: { useMGet: true }): void;
function test(client: GetClient, options: { useMGet: false }): void;
function test(client: GetClient | MGetClient, options: { useMGet: boolean } = { useMGet: true }): void  {
    return undefined;
}

test({ mget: () => {}}); // ok 
test({ get: () => {}}, { useMGet: false }); // ok
test({ mget: () => {}}, { useMGet: true }); // ok

test({ get: () => {}}); // type error
test({ get: () => {}}, { useMGet: true }); // type error
test({ mget: () => {}}, { useMGet: false }); // type error
```

Reply:
*nod* yeah, that definitely works, but I think I'm going to go with this way (which provides type checking even if you're being more dynamic, whereas yours only provides full type checking if you're using literal `true` or `false`).